src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/FormLoginLdapFactory.php

```diff
@@ -26,15 +26,33 @@ class FormLoginLdapFactory extends FormLoginFactory
 {
     protected function createAuthProvider(ContainerBuilder $container, $id, $config, $userProviderId)
     {
-        $provider = 'security.authentication.provider.ldap_bind.'.$id;
-        $container
-            ->setDefinition($provider, new ChildDefinition('security.authentication.provider.ldap_bind'))
-            ->replaceArgument(0, new Reference($userProviderId))
-            ->replaceArgument(1, new Reference('security.user_checker.'.$id))
-            ->replaceArgument(2, $id)
-            ->replaceArgument(3, new Reference($config['service']))
-            ->replaceArgument(4, $config['dn_string'])
-        ;
+        $method = empty($config['query_string']) ? 'bind' : 'query';
+        
+        $provider = 'security.authentication.provider.ldap_'.$method.'.'.$id;
+
+        switch($method) {
```

Comment:
there is probably a more elegant way to do this without duplicating so much code